stl/inc/xutility

```diff
@@ -260,7 +260,8 @@ struct pointer_traits {
 
     using _Reftype = conditional_t<is_void_v<element_type>, char, element_type>&;
 
-    _NODISCARD static pointer pointer_to(_Reftype _Val) noexcept(noexcept(_Ty::pointer_to(_Val))) /* strengthened */ {
+    _NODISCARD static _CONSTEXPR20 pointer pointer_to(_Reftype _Val) noexcept(

```

Comment:
This is speculatively implementing LWG-3454. Perhaps we should add a comment.

Reply:
I added a comment.
clang 15 implemented `constexpr vector` .
And without the `_CONSTEXPR20` we have 85 failed tests:
```
Failed Tests (85):
  libc++ :: std/containers/sequences/vector/access.pass.cpp:0
  libc++ :: std/containers/sequences/vector/access.pass.cpp:1
  libc++ :: std/containers/sequences/vector/contiguous.pass.cpp:0
  libc++ :: std/containers/sequences/vector/contiguous.pass.cpp:1
  libc++ :: std/containers/sequences/vector/iterators.pass.cpp:1
  libc++ :: std/containers/sequences/vector/reverse_iterators.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.capacity/capacity.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.capacity/capacity.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.capacity/empty.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.capacity/empty.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.capacity/reserve.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.capacity/reserve.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.capacity/resize_size.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.capacity/resize_size.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.capacity/resize_size_value.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.capacity/resize_size_value.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.capacity/shrink_to_fit.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.capacity/shrink_to_fit.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.capacity/size.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.capacity/size.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.capacity/swap.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.capacity/swap.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.cons/assign_initializer_list.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.cons/assign_initializer_list.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.cons/assign_move.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.cons/assign_move.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.cons/assign_size_value.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.cons/assign_size_value.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.cons/construct_default.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.cons/construct_default.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.cons/construct_iter_iter.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.cons/construct_iter_iter_alloc.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.cons/construct_size.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.cons/construct_size_value.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.cons/construct_size_value_alloc.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.cons/copy.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.cons/copy.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.cons/copy_alloc.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.cons/copy_alloc.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.cons/initializer_list.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.cons/initializer_list.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.cons/initializer_list_alloc.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.cons/initializer_list_alloc.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.cons/move.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.cons/move.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.cons/move_alloc.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.cons/move_alloc.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.cons/op_equal_initializer_list.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.cons/op_equal_initializer_list.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.data/data.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.data/data.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.data/data_const.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.data/data_const.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.erasure/erase.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.erasure/erase_if.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.modifiers/clear.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.modifiers/clear.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.modifiers/emplace.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.modifiers/emplace.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.modifiers/emplace_back.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.modifiers/emplace_back.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.modifiers/emplace_extra.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.modifiers/emplace_extra.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.modifiers/erase_iter.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.modifiers/erase_iter.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.modifiers/erase_iter_iter.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.modifiers/erase_iter_iter.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.modifiers/insert_iter_initializer_list.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.modifiers/insert_iter_initializer_list.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.modifiers/insert_iter_iter_iter.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.modifiers/insert_iter_iter_iter.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.modifiers/insert_iter_lvalue.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.modifiers/insert_iter_lvalue.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.modifiers/insert_iter_rvalue.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.modifiers/insert_iter_rvalue.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.modifiers/insert_iter_size_value.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.modifiers/insert_iter_size_value.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.modifiers/pop_back.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.modifiers/pop_back.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.modifiers/push_back.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.modifiers/push_back.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.modifiers/push_back_rvalue.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.modifiers/push_back_rvalue.pass.cpp:1
  libc++ :: std/containers/sequences/vector/vector.special/swap.pass.cpp:0
  libc++ :: std/containers/sequences/vector/vector.special/swap.pass.cpp:1
```

Error example:
```
C:\Dev\STL\llvm-project\libcxx\test\std\containers\sequences\vector\vector.modifiers\clear.pass.cpp(50,19): error: static_assert expression is not an integral constant expression
    static_assert(tests());
                  ^~~~~~~
C:\Dev\STL\out\build\x64\out\inc\xmemory(986,40): note: non-constexpr function 'pointer_to' cannot be used in a constant expression
        _Alloc_traits::deallocate(_Al, _Ptr_traits::pointer_to(*_Ptr), 1);
                                       ^
C:\Dev\STL\out\build\x64\out\inc\xmemory(995,5): note: in call to '_Deallocate_plain(static_cast<_Rebind_alloc_t<std::vector<int, min_allocator<int>>::_Alty, std::_Container_proxy>>(this->_Getal()), &{*new std::_Container_proxy[1]#2}[0])'
    _Deallocate_plain(_Al, _Ptr);
    ^
C:\Dev\STL\out\build\x64\out\inc\vector(813,9): note: in call to '_Delete_plain_internal(static_cast<_Rebind_alloc_t<std::vector<int, min_allocator<int>>::_Alty, std::_Container_proxy>>(this->_Getal()), &{*new std::_Container_proxy[1]#2}[0])'
        _Delete_plain_internal(_Alproxy, _STD exchange(_Mypair._Myval2._Myproxy, nullptr));
        ^
C:\Dev\STL\llvm-project\libcxx\test\std\containers\sequences\vector\vector.modifiers\clear.pass.cpp(34,42): note: in call to '&c->~vector()'
    std::vector<int, min_allocator<int>> c(a, a+3);
                                         ^
C:\Dev\STL\llvm-project\libcxx\test\std\containers\sequences\vector\vector.modifiers\clear.pass.cpp(50,19): note: in call to 'tests()'
    static_assert(tests());
                  ^
C:\Dev\STL\out\build\x64\out\inc\xutility(263,31): note: declared here
    _NODISCARD static pointer pointer_to(_Reftype _Val) noexcept(
                              ^
1 error generated.
```
So maybe we have bugs in our `vector` implementation...

Reply:
A similar failure involving `pointer_to` and `string` is what motivated me to amend LWG-3454 in the first place. The Standard doesn't make that overload of `pointer_to` `constexpr`, so there's no way to manufacture a fancy pointer value at compile time for a fancy pointer type that uses the primary `pointer_traits` template. We could (by breaking ABI) avoid the problem for `vector` by storing fancy pointers instead of raw pointers for allocated proxies, but for `string` you effectively must choose between fancy pointers and SSO.

I favor speculative implementation of LWG-3454, but we could theoretically fix the tests instead by explicitly specializing `pointer_traits` for the fancy pointer types in question, making `pointer_to` `constexpr` in the process.